File: sdks/python/tests/fixtures/protocol_worker.py
"""Independent author fixture for exclusive and concurrent wire scenarios."""

import asyncio
import io
import os
import time
from weaveport_sdk import PluginApplication

app = PluginApplication(
    concurrent_calls=os.environ["WEAVEPORT_TEST_CONCURRENT"] == "true"
)


@app.function("work")
async def work(value, context):
    await asyncio.sleep(value.get("delay", 0))
    if value.get("callback"):
        return await context.call_host("echo", context.tenant)
    return context.tenant


@app.function("sync")
def sync(value, context):
    time.sleep(value.get("delay", 0))
    if value.get("callback"):
        return context.call_host_sync("echo", context.tenant)
    return context.tenant


@app.function("cleanup")
async def cleanup(value, context):
    async def close():
        await asyncio.sleep(value.get("delay", 0))
        if value.get("fail"):
            raise RuntimeError("cleanup failed")

    context.on_close(close)
    return context.tenant


@app.function("invalid")
async def invalid(value, context):
    return object()


@app.stream("brokenClose")
async def broken_close(value, context):
    try:
        yield 1
        await asyncio.sleep(0.2)
        yield 2
    finally:
        raise RuntimeError("iterator close failed")


@app.stream("immediateCallback")
async def immediate_callback(value, context):
    if value.get("before"):
        yield await context.call_host("echo", 0)
    yield 1
    yield await context.call_host("echo", 2)


@app.stream("items")
async def items(value, context):
    yield 1
    await asyncio.sleep(0.12)
    yield await context.call_host("echo", 2)


@app.source("bytes")
def source(value, context):
    return io.BytesIO(b"hello")


app.run()

File: sdks/python/tests/fixtures/protocol_worker.mjs
// Independent author fixture for exclusive and concurrent wire scenarios.
const { PluginApplication } = await import(process.env.WEAVEPORT_TEST_SDK);
const app = new PluginApplication('1', {
    concurrentCalls: process.env.WEAVEPORT_TEST_CONCURRENT === 'true',
});
app.function('work', async (value, context, signal) => {
    await new Promise((resolve) => setTimeout(resolve, (value.delay ?? 0) * 1000));
    if (value.callback) {
        return await context.callHost('echo', context.tenant);
    }
    return context.tenant;
});
app.function('cleanup', async (value, context) => {
    context.onClose(async () => {
        await new Promise((resolve) => setTimeout(resolve, (value.delay ?? 0) * 1000));
        if (value.fail) {
            throw new Error('cleanup failed');
        }
    });
    return context.tenant;
});
app.function('invalid', async () => 1n);
app.stream('brokenClose', async function* () {
    try {
        yield 1;
        await new Promise((resolve) => setTimeout(resolve, 100));
        yield 2;
    } finally {
        throw new Error('iterator close failed');
    }
});
app.stream('immediateCallback', async function* (value, context) {
    if (value.before) {
        yield await context.callHost('echo', 0);
    }
    yield 1;
    yield await context.callHost('echo', 2);
});
app.stream('items', async function* (value, context) {
    yield 1;
    await new Promise((resolve) => setTimeout(resolve, 120));
    yield await context.callHost('echo', 2);
});
app.source('bytes', () => {
    let sent = false;
    return {
        read() {
            if (sent) {
                return new Uint8Array();
            }
            sent = true;
            return Buffer.from('hello');
        },
        close() {},
    };
});
await app.run();
